### ISSUE TITLE
Dict preferences edited with PlistBuddy not read by Aerial.

### ISSUE THREAD

**Issue by kdeldycke** [JohnCoates/Aerial#976]
#### General troubleshooting tips

Before logging an issue please check that:
- [x] You have the latest version installed (There may be a beta version that fixes your issue), see here for the latest releases and bug fixes: https://github.com/JohnCoates/Aerial/releases
- [x] Your issue isn't already mentioned in our [issues](https://github.com/JohnCoates/Aerial/issues). You may find a workaround there or a similar request already made.
- [x] Your problem isn't mentioned in the [troubleshooting section](https://github.com/JohnCoates/Aerial#troubleshooting).

If none of this fixes your issue, tell us about the problem you are experiencing or the feature you'd like to request.

#### Required information
In order to help us sort your issue, we ask that you provide the following information:
- [x] Mac model: MacBook Air (Retina, 13-inch, 2018)
- [x] macOS version: Catalina 10.15.4 (19E287)
- [x] Monitor setup: 1 additional 4k Samsung monitor.

If appropriate, please enable `Debug mode` and `Log to disk` in `Advanced` tab and replicate your bug, then attach the `AerialLog.txt` file (You can access this file through the Advanced tab).

#### Description of issue / Feature request

Aerial preferences in Catalina are saved in `~/Library/Containers/com.apple.ScreenSaver.Engine.legacyScreenSaver/Data/Library/Preferences/ByHost/com.JohnCoates.Aerial.<UUID>.plist`:

```
$ /usr/libexec/PlistBuddy -c "Print" ~/Library/Containers/com.apple.ScreenSaver.Engine.legacyScreenSaver/Data/Library/Preferences/ByHost/com.JohnCoates.Aerial.<UUID>.plist

Dict {
    LayerLocation = {"displays":0,"fontName":"Helvetica Neue Medium","time":0,"isEnabled":true,"corner":7,"fontSize":28}
    layers = ["message","clock","location","battery","countdown","timer","date"]
    fadeMode = 0
    debugMode = false
    LayerClock = {"isEnabled":true,"corner":3,"displays":1,"fontSize":50,"fontName":"Helvetica Neue Medium","showSeconds":false,"hideAmPm":true}
    logToDisk = false
    intOnBatteryMode = 2
    neverStreamVideos = true
    updateWhileSaverMode = false
    newVideosMode = 1
    aspectMode = 0
    timeMode = 3
}
```

These preferences works fine and have been produced by tweaking options manually in Aerial preferences panel. Notice how `LayerLocation` entry is written.

Now, let's say I way to programmatically change these preferences with `PlistBuddy`:
```
$ /usr/libexec/PlistBuddy \
    -c "Delete  :LayerLocation" \
    -c "Add     :LayerLocation:displays integer 1" \
    -c "Add     :LayerLocation:time     integer 1" \
    ~/Library/Containers/com.apple.ScreenSaver.Engine.legacyScreenSaver/Data/Library/Preferences/ByHost/com.JohnCoates.Aerial.<UUID>.plist
```

See how the `LayerLocation` entry is now rendered:
```
$ /usr/libexec/PlistBuddy -c "Print" ~/Library/Containers/com.apple.ScreenSaver.Engine.legacyScreenSaver/Data/Library/Preferences/ByHost/com.JohnCoates.Aerial.<UUID>.plist

Dict {
    LayerLocation = Dict {
        displays = 1
        time = 1
    }
    layers = ["message","clock","location","battery","countdown","timer","date"]
    (...)
}
```

As you can see it looks like the preferences produced by Aerial have dictionnaries serialized as linear strings while plist utilities produces real Dict data structures.

This might explain why, after my PlistBuddy edit above the preferences are not applied by Aerial.

**Comment by kdeldycke** [comment]
Some related issues to help in debugging: #917 & #958.

**Comment by kdeldycke** [comment]
For the context, I uncovered this issue while attempting to automate the setting of all Aerial preferences in my dotfiles at: https://github.com/kdeldycke/dotfiles/commit/cb6a6f7f73f80fe415f2dd9f0ccf5399d74a8965

**Comment by glouel** [comment]
Hi @kdeldycke !

So, yeah, the settings for the Infos (what's called layers in the code, the various layers of infos that are overlayed on videos) are serialized structs in Swift.

**Comment by kdeldycke** [comment]
OK, thanks @glouel for clarifying the situation. I'll try to format the value into serialized strings and see how Aerial reacts.

**Comment by kdeldycke** [comment]
I just tried to hack preferences and use hard-serialized strings like so:

```
$ /usr/libexec/PlistBuddy -c 'Delete :LayerLocation' -c 'Add :LayerLocation string {\"displays\":0,\"fontName\":\"Helvetica Neue Medium\",\"time\":0,\"isEnabled\":true,\"corner\":7,\"fontSize\":12}' ~/Library/Containers/com.apple.ScreenSaver.Engine.legacyScreenSaver/Data/Library/Preferences/ByHost/com.JohnCoates.Aerial.<UUID>.plist
```

Which results in:
```
❯ /usr/libexec/PlistBuddy -c "Print" ~/Library/Containers/com.apple.ScreenSaver.Engine.legacyScreenSaver/Data/Library/Preferences/ByHost/com.JohnCoates.Aerial.<UUID>.plist
Dict {
    LayerLocation = {"displays":0,"fontName":"Helvetica Neue Medium","time":0,"isEnabled":true,"corner":7,"fontSize":12}
    (...)
}
```

But this doesn't work at all. While preferences are kept persisting in the plist file, Aerial doesn't pick up the new preferences, even after a reboot.

And for completeness, even a sub set of value doesn't work either.

**Comment by glouel** [comment]
> I just tried to hack preferences and use hard-serialized strings like so:
> (...)
> ```
> ❯ /usr/libexec/PlistBuddy -c "Print" ~/Library/Containers/com.apple.ScreenSaver.Engine.legacyScreenSaver/Data/Library/Preferences/ByHost/com.JohnCoates.Aerial.<UUID>.plist
> Dict {
>     LayerLocation = {"displays":0,"fontName":"Helvetica Neue Medium","time":0,"isEnabled":true,"corner":7,"fontSize":12}
>     (...)
> }
> ```

Hey again

So yeah that definitely won't work. It's not a string, it's a Data blob, and at the very least, the struct is ordered so you'd need to have the parameters in the correct order. Here's the struct : 

```
    struct Location: CommonInfo, Codable {
        var isEnabled: Bool
        var fontName: String
        var fontSize: Double
        var corner: InfoCorner
        var displays: InfoDisplays
        var time: InfoTime
    }
```

The last 3 types are Enums that map to a Codable Int, so just assume it's an int. Here's the default settings: 

```
Location(isEnabled: true,
         fontName: "Helvetica Neue Medium",
         fontSize: 28,
         corner: .random,
         displays: .allDisplays,
         time: .always)
```

With `corner = 7, displays = 0, time = 0`

But I think you are massively overcomplicating things. Here's what I would do, configure it once, take the value from the Data blob (it's a Data type in the plist, not a string), and add that to your config. I understand it's inelegant but it should work. I may move the settings at some point too as mentioned previously, hopefully to something more human readable.

**Comment by glouel** [comment]
Quick important update (sorry I don't touch that part of the code much so I forgot how it worked), it's supposed to be json encoded, so I think you kinda had the right idea. I think the problem is that somehow I'm storing the string as data. Assuming the right order for the parameters your thing might just work. 

But I might want to look at why I'm not just storing the string as a string as I intended 🙈

Stay tuned I'll try to have a look.

**Comment by glouel** [comment]
@kdeldycke well I fixed it, while keeping retro compatibility with the old format. So now I'm storing those structs as a JSON strings as I had originally intended, which will make your edits easier. Here's how it looks : 

```
$ defaults read ~/Library/Preferences/ByHost/com.JohnCoates.Aerial.plist LayerLocation
{
  "corner" : 0,
  "displays" : 0,
  "fontName" : "Helvetica Neue Medium",
  "fontSize" : 28,
  "isEnabled" : true,
  "time" : 0
}
```

That should be a lot more hackable for you ;) You'll need the latest beta but the changes will remain in the upcoming releases. Let me know how it works for you : 

https://github.com/JohnCoates/Aerial/releases/download/v1.8.3beta7/Aerial.saver.zip

**Comment by glouel** [comment]
Didn't hear back from you so closing this, but as far as I know this is now fixed. Open it up if you have an issue again.

**Comment by ruzickap** [comment]
Hello guys.

I hit the same issue and I can not change the parameters inside `Layer*` variables.

I also looked at https://github.com/JohnCoates/Aerial/issues/976#issuecomment-619547750, which looks promising, but there is no such file (using Aerial 1.9.1) [I tried to find it using `find ~/Library/Preferences/ -name "*Aerial*"`].

If you have a solution how to "fully" configure Aerial from command line, please update this ticket with working example.

Thank you

**Comment by glouel** [comment]
Hi @ruzickap 

So a couple of things. 

First, the plist is located either at :

`~/Library/Preferences/ByHost/com.JohnCoates.Aerial.{UUID}.plist` (before Catalina)
`~/Library/Containers/com.apple.ScreenSaver.Engine.legacyScreenSaver/Data/Library/Preferences/ByHost/com.JohnCoates.Aerial.{UUID}.plist` (starting with Catalina, because of containerization) 

Do NOT edit the plist that may be a level below with the same name. 

Next, I don't use Plistbuddy but here's how to configure a LayerWeather to be enabled in Amsterdam with basic defaults write. If you adapt it to your tool, feel free to post the command too : 

Pre Catalina (I tested this one) : 
```
defaults write ~/Library/Preferences/ByHost/com.JohnCoates.Aerial.plist LayerWeather -string "{\"corner\" : 1, \"degree\" : 0, \"displays\" : 0, \"fontName\" : \"Helvetica Neue Medium\", \"fontSize\" : 40, \"icons\" : 0,  \"isEnabled\" : true, \"locationMode\" : 1,  \"locationString\" : \"Amsterdam, NL\"}"
```

Post Catalina : 

```
defaults write ~/Library/Containers/com.apple.ScreenSaver.Engine.legacyScreenSaver/Data/Library/Preferences/ByHost/com.JohnCoates.Aerial.plist LayerWeather -string "{\"corner\" : 1, \"degree\" : 0, \"displays\" : 0, \"fontName\" : \"Helvetica Neue Medium\", \"fontSize\" : 40, \"icons\" : 0,  \"isEnabled\" : true, \"locationMode\" : 1,  \"locationString\" : \"Amsterdam, NL\"}"
```

With defaults write you don't put the UUID but it's there in the filename.

**Comment by ruzickap** [comment]
Thanks a lot :-)
Works like a charm now...

**Comment by kdeldycke** [comment]
Sorry @glouel for the delayed answers, but thanks to @ruzickap troubleshooting, I finally got to the bottom of it.

TL;DR:
* Use `defaults` instead of `PlistBuddy`. The later [doesn't send the necessary system signals](https://scriptingosx.com/2016/11/editing-property-lists/) to have running process re-read and update the modified configuration.
* Make sure all keys in layer's configuration are set. Aerial will not pickup the sub-dictionary diff to merge to the defaults value. The serialized JSON string will be taken as a whole.

All in all, here are my final (working) config: https://github.com/kdeldycke/dotfiles/blob/f9e04b018a8b77f9aecd45f7e108ea934c544a86/macos-config.sh#L501-L597
